File: example/moon/src/main/java/com/moon/models/Address.java
package com.moon.models;


public record Address(String street, String city, String zipCode, String country, String supplement) {

    public static class Builder {
        private String street;
        private String city;
        private String zipCode;
        private String country;
        private String supplement;

        public Builder() {}

        public Builder street(String street) {
            this.street = street;
            return this;
        }

        public Builder city(String city) {
            this.city = city;
            return this;
        }

        public Builder zipCode(String zipCode) {
            this.zipCode = zipCode;
            return this;
        }

        public Builder country(String country) {
            this.country = country;
            return this;
        }

        public Builder supplement(String supplement) {
            this.supplement = supplement;
            return this;
        }

        public Address build() {

            return new Address(
                this.street,
                this.city,
                this.zipCode,
                this.country,
                this.supplement
            );
        }

    }
}

File: kotlin/address/Address.kt
package kotlin.address

data class Address(
    val street: String? = null,
    val city: String? = null,
    val zipCode: String? = null,
    val country: String? = null,
    val supplement: String? = null
)

fun main() {
    val address = Address(
        street = "Rue de la Paix",
        city = "Paris",
        zipCode = "75000",
        country = "France",
        supplement = "Appartement 123"
    )

    val javaAddress  = AddressJ.Builder()
        .street("Rue de la Paix")
        .city("Paris")
        .zipCode("75000")
        .country("France")
        .supplement("Appartement 123")
        .build()

}